### ISSUE TITLE
more failures in county-specific queries

### ISSUE THREAD

**Issue by lindsaycarr** [USGS-R/geoknife#282]
Want to query for Taylor County in FL. Using webgeom and query, I determined that `Taylor County` was valid value I could query from `geom = 'derivative:US_Counties'` and `attribute = 'COUNTY'`. However, when you use `values = 'Taylor County'`, the resulting data comes back with data for 6 counties, none of which are Taylor.

```r
wg_counties <- webgeom(geom = 'derivative:US_Counties', attribute = 'COUNTY')
counties <- query(wg_counties , 'values')
wg_states <- webgeom(geom = 'derivative:US_Counties', attribute = 'STATE')
states <- query(wg_states, 'values')
county_info <- data.frame(state = states, county = counties)

stencil <- webgeom(geom = 'derivative:US_Counties', 
                   attribute = 'COUNTY',
                   values = 'Taylor County')

fabric <- webdata(url = 'http://cida.usgs.gov/thredds/dodsC/stageiv_combined', 
                  variables = "Total_precipitation_surface_1_Hour_Accumulation", 
                  times = c(as.POSIXct("2016-06-06 05:00:00"), 
                            as.POSIXct("2016-06-07 05:00:00")))

job <- geoknife(stencil, fabric, wait = TRUE)
precipData <- result(job)
```

![image](https://cloud.githubusercontent.com/assets/13220910/15868858/3884a804-2caf-11e6-9100-539ee55a2717.png)

**Comment by aappling-usgs** [comment]
digging in. 
```r
> stencil@GML_IDs
[1] "US_Counties.1807" "US_Counties.2782" "US_Counties.3128" "US_Counties.3685" "US_Counties.4782"
[6] "US_Counties.4823"
```

using breakpoint to stop at geoknife() line 69, i see these IDs again in the XML:
```
Browse[2]> geojob
An object of class "geojob":
url: http://cida.usgs.gov/gdp/process/WebProcessingService 
xml:
 <wps:Execute xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:ogc="http://www.opengis.net/ogc" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" service="WPS" version="1.0.0" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsExecute_request.xsd">
...
    <wps:Input>
      <ows:Identifier>FEATURE_COLLECTION</ows:Identifier>
      <wps:Reference xlink:href="http://cida.usgs.gov/gdp/geoserver/wfs">
        <wps:Body>
          <wfs:GetFeature xmlns:wfs="http://www.opengis.net/wfs" xmlns:gml="http://www.opengis.net/gml" service="WFS" version="1.1.0" outputFormat="text/xml; subtype=gml/3.1.1">
            <wfs:Query typeName="derivative:US_Counties">
              <wfs:PropertyName>the_geom</wfs:PropertyName>
              <wfs:PropertyName>COUNTY</wfs:PropertyName>
              <ogc:Filter>
                <ogc:GmlObjectId gml:id="US_Counties.1807"/>
                <ogc:GmlObjectId gml:id="US_Counties.2782"/>
                <ogc:GmlObjectId gml:id="US_Counties.3128"/>
                <ogc:GmlObjectId gml:id="US_Counties.3685"/>
                <ogc:GmlObjectId gml:id="US_Counties.4782"/>
                <ogc:GmlObjectId gml:id="US_Counties.4823"/>
              </ogc:Filter>
            </wfs:Query>
          </wfs:GetFeature>
        </wps:Body>
      </wps:Reference>
    </wps:Input>
...
```
if these are the wrong IDs, then the bug is somewhere in `fetchGML_IDs`. if they're the right IDs, then probably the bug is with the wfs when it translates IDs into shapefiles/similar.

**Comment by aappling-usgs** [comment]
redid code (PR in a minute). new results:
```
> stencil@GML_IDs
[1] "US_Counties.1813" "US_Counties.2794" "US_Counties.3140" "US_Counties.3697" "US_Counties.4794"
[6] "US_Counties.4835"
```
(good)

```
> precipData
              DateTime Taylor County                                        variable statistic
1  2016-06-06 05:00:00  6.151087e-03 Total_precipitation_surface_1_Hour_Accumulation      MEAN
2  2016-06-06 06:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
3  2016-06-06 07:00:00  9.955345e-04 Total_precipitation_surface_1_Hour_Accumulation      MEAN
4  2016-06-06 08:00:00  6.281916e-05 Total_precipitation_surface_1_Hour_Accumulation      MEAN
5  2016-06-06 09:00:00  3.986425e-03 Total_precipitation_surface_1_Hour_Accumulation      MEAN
6  2016-06-06 10:00:00  8.442339e-03 Total_precipitation_surface_1_Hour_Accumulation      MEAN
7  2016-06-06 11:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
8  2016-06-06 12:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
9  2016-06-06 13:00:00  2.816793e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
10 2016-06-06 14:00:00  9.593131e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
11 2016-06-06 15:00:00  9.207123e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
12 2016-06-06 16:00:00  1.096228e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
13 2016-06-06 17:00:00  5.527621e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
14 2016-06-06 18:00:00  1.436294e-01 Total_precipitation_surface_1_Hour_Accumulation      MEAN
15 2016-06-06 19:00:00  3.920711e-01 Total_precipitation_surface_1_Hour_Accumulation      MEAN
16 2016-06-06 20:00:00  2.230885e-01 Total_precipitation_surface_1_Hour_Accumulation      MEAN
17 2016-06-06 21:00:00  1.234502e-01 Total_precipitation_surface_1_Hour_Accumulation      MEAN
18 2016-06-06 22:00:00  7.999857e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
19 2016-06-06 23:00:00  2.730543e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
20 2016-06-07 00:00:00  1.488369e-02 Total_precipitation_surface_1_Hour_Accumulation      MEAN
21 2016-06-07 01:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
22 2016-06-07 02:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
23 2016-06-07 03:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
24 2016-06-07 04:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
25 2016-06-07 05:00:00  0.000000e+00 Total_precipitation_surface_1_Hour_Accumulation      MEAN
```
(unexpected - i thought we'd get 6 columns, one for each of the 6 Taylor Counties. @jread-usgs, which behavior is desired, and where is that behavior communicated in the documentation?)

**Comment by lindsaycarr** [comment]
Wow, awesome! Thanks for the quick fix. Any thoughts on how I can be specific about wanting the Taylor County from Florida?

**Comment by jread-usgs** [comment]
So...the averaging to a single Taylor County is because GDP will combine all values with the same name (so all of the `COUNTY` values here are the same). To avoid it, you can choose a different attribute that has unique values for all polygons for the counties. I think FIP does this. 

The bug though w/ the wrong counties coming back is a bug, and I hope it is related to the query returning more that 5000 results (that is the cap).

**Comment by lindsaycarr** [comment]
Tried FIPS, but it comes back with an error:

```r
stencil <- webgeom(geom = 'derivative:US_Counties',
                   attribute = 'FIPS',
                   values = '12123')

Error in (function (cl, name, valueClass)  : 
  assignment of an object of class “NULL” is not valid for @‘GML_IDs’ in an object of class “webgeom”; is(value, "character") is not TRUE
```

**Comment by aappling-usgs** [comment]
oh, right - i think what we learned was that the only values you can search on for geom=US_Counties are the county names. this fits with the xml i was just inspecting - the only info in the xml in `fetchGML_IDs` is the County Name and the IDs of form "US_Counties.1813"

**Comment by jiwalker-usgs** [comment]
I think the derivative versions of those shapefiles has some of the attributes stripped out.  sample:US_Counties might be better (if it exists).

**Comment by aappling-usgs** [comment]
i'm taking back what i just wrote because i just realized that you can change what xml comes back by setting a different `attribute`, so the problem @lindsaycarr just pointed out happens somewhere after that (and i'm looking into it).

not sure what you mean, @jiwalker-usgs - could you please elaborate?

**Comment by jiwalker-usgs** [comment]
Nevermind, sample:US_counties does not exist, so derivative is the right workspace and FIPS is there.  Looks like you are on the right path.

**Comment by aappling-usgs** [comment]
@lindsaycarr @jread-usgs - ok, so the problem with searching for FIPS=12123 is that this FIPS code is not among the 5000 elements returned by the query (because the query only asks for 5000, not more, and there are >5000 counties in the US). see #277, which would solve this by asking the WFS to do the query locally before sending back just the matching elements.

**Comment by jread-usgs** [comment]
yes, that is exactly the issue coming up here.

**Comment by aappling-usgs** [comment]
yep, definitely a bug. but @lindsaycarr , the good news is that searching by FIPS is possible as long as you pick 1+ of the lucky returned elements:

```
> stencil <- webgeom(geom = 'derivative:US_Counties', 
+                    attribute = 'FIPS',
+                    values = c('05015','21141'))
> stencil@GML_IDs
[1] "US_Counties.3813" "US_Counties.3968"
> job <- geoknife(stencil, fabric, wait = TRUE)
> check(job)
$status
[1] "Process successful"

$URL
[1] "http://cida.usgs.gov:80/gdp/process/RetrieveResultServlet?id=5c19ff4e-97f2-4a17-951d-1c99b56abf3dOUTPUT"

$statusType
[1] "ProcessSucceeded"

> precipData <- result(job)
> head(precipData)
             DateTime 05015 21141                                        variable statistic
1 2016-06-06 05:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
2 2016-06-06 06:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
3 2016-06-06 07:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
4 2016-06-06 08:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
5 2016-06-06 09:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
6 2016-06-06 10:00:00     0     0 Total_precipitation_surface_1_Hour_Accumulation      MEAN
```

**Comment by jread-usgs** [comment]
How should we proceed @aappling-usgs ? is #277 a priority or do we want a helper patch in the meantime? I think you said you were working on a PR

**Comment by aappling-usgs** [comment]
i am about to submit a PR for the ID matching bug (with new documentation for multiple counties of same name), but i think you can do #277 faster than I can. after i submit my PR, you have any time to do the POST solution, and/or want to pair program it with me?

**Comment by lindsaycarr** [comment]
Ok, so I just can't use that FIPS because it's greater than the 5000th county listed...but I need specific counties because I'm picking ones that were in the storm's path. How quickly could there be a work-around for this?

**Comment by jread-usgs** [comment]
quickly, but not so quickly to get that fix to CRAN if you are wanted to showcase something that others could do. 

The quick fix is to change the query max on https://github.com/USGS-R/geoknife/blob/master/R/values-webgeom.R#L53 from 5000 to 20000 or something larger than the number of counties. 

The better fix is to implement #277

**Comment by lindsaycarr** [comment]
Can we just specify they should `devtools:install_github('USGS-R/geoknife')` in our code?
